reactos/drivers/usb/usbhub_new/ioctl.c

```diff
@@ -0,0 +1,1406 @@
+}
+
+NTSTATUS
+NTAPI
+USBH_PortIdleNotificationRequest(IN PUSBHUB_PORT_PDO_EXTENSION PortExtension,
+                                 IN PIRP Irp)
+{
+    PUSBHUB_FDO_EXTENSION HubExtension;
+    PIO_STACK_LOCATION IoStack;
+    PUSB_IDLE_CALLBACK_INFO IdleCallbackInfo;
+    NTSTATUS Status;
+    KIRQL Irql;
+
+    DPRINT("USBH_PortIdleNotificationRequest ... \n");
+
+    HubExtension = PortExtension->HubExtension;
+
+    IoAcquireCancelSpinLock(&Irql);
+
+    if (PortExtension->IdleNotificationIrp)
+    {
+        IoReleaseCancelSpinLock(Irql);
+        Irp->IoStatus.Status = STATUS_DEVICE_BUSY;
+        IoCompleteRequest(Irp, IO_NO_INCREMENT);
+        return STATUS_DEVICE_BUSY;
+    }
+
+    IoStack = IoGetCurrentIrpStackLocation(Irp);
+    IdleCallbackInfo = IoStack->Parameters.DeviceIoControl.Type3InputBuffer;
+
+    if (!IdleCallbackInfo || !IdleCallbackInfo->IdleCallback)
+    {
+        IoReleaseCancelSpinLock(Irql);
+
+        Status = STATUS_NO_CALLBACK_ACTIVE;
+        Irp->IoStatus.Status = Status;
+        IoCompleteRequest(Irp, IO_NO_INCREMENT);
+
+        return Status;
+    }
+
+    IoSetCancelRoutine(Irp, USBH_PortIdleNotificationCancelRoutine);
+
+    if (Irp->Cancel)
+    {
+        if (IoSetCancelRoutine(Irp, NULL))
+        {
+            IoReleaseCancelSpinLock(Irql);
+            Status = STATUS_CANCELLED;
+            Irp->IoStatus.Status = STATUS_CANCELLED;
+            IoCompleteRequest(Irp, IO_NO_INCREMENT);
+        }
+        else
+        {
+            IoMarkIrpPending(Irp);
+            IoReleaseCancelSpinLock(Irql);
+            Status = STATUS_PENDING;
+        }
+    }
+    else
+    {
+        PortExtension->PortPdoFlags |= USBHUB_PDO_FLAG_IDLE_NOTIFICATION;
+
+        PortExtension->IdleNotificationIrp = Irp;
+        IoMarkIrpPending(Irp);
+
+        IoReleaseCancelSpinLock(Irql);
+        Status = STATUS_PENDING;
+
+        DPRINT("USBH_PortIdleNotificationRequest: IdleNotificationIrp - %p\n",
+               PortExtension->IdleNotificationIrp);
+
+        USBH_CheckIdleDeferred(HubExtension);
+    }
+
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+USBH_IoctlGetNodeName(IN PUSBHUB_FDO_EXTENSION HubExtension,
+                      IN PIRP Irp)
+{
+    PUSB_NODE_CONNECTION_NAME ConnectionName;
+    PDEVICE_OBJECT PortDevice;
+    PUSBHUB_PORT_PDO_EXTENSION PortExtension;
+    ULONG LengthSkip;
+    PWCHAR Buffer;
+    ULONG BufferLength;
+    PWCHAR BufferEnd;
+    ULONG LengthReturned;
+    ULONG LengthName;
+    ULONG Length;
+    NTSTATUS Status;
+    PIO_STACK_LOCATION IoStack;
+    ULONG_PTR Information;
+
+    DPRINT("USBH_IoctlGetNodeName ... \n");
+
+    Status = STATUS_SUCCESS;
+
+    ConnectionName = Irp->AssociatedIrp.SystemBuffer;
+
+    IoStack = IoGetCurrentIrpStackLocation(Irp);
+    Length = IoStack->Parameters.DeviceIoControl.OutputBufferLength;
+
+    if (Length < sizeof(USB_NODE_CONNECTION_NAME))
+    {
+        Status = STATUS_BUFFER_TOO_SMALL;
+        goto Exit;
```

Comment:
ioctl.c:579: In this case (and below in line 586), `Information` will be used uninitialized.